src/main/java/io/cdap/plugin/hubspot/sink/batch/HubspotSinkHelper.java

```diff
@@ -0,0 +1,135 @@
+
+import org.apache.http.client.utils.URIBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.ProtocolException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Helper class to incorporate Hubspot sink api interaction
+ */
+public class HubspotSinkHelper {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HubspotSinkHelper.class);
+
+  public static final int MAX_TRIES = 3;
+
+  public static final String DEFAULT_API_SERVER_URL = "https://api.hubapi.com";
+
+  public void executeHTTPService(String input, SinkHubspotConfig config) throws Exception {
+
+    int responseCode;
+    int retries = 0;
+    Exception exception = null;
+    do {
+      HttpURLConnection conn = null;
+      Map<String, String> headers = new HashMap<>();
+      try {
+         URIBuilder b = new URIBuilder(getSinkEndpoint(config));
+         if (config.apiKey != null) {
+           b.addParameter("hapikey", config.apiKey);
+         }
+        URL url = new URL(b.build().toString());
+        conn = (HttpURLConnection) url.openConnection();
+        conn.setRequestMethod("POST");
+        for (Map.Entry<String, String> propertyEntry : headers.entrySet()) {
+          conn.addRequestProperty(propertyEntry.getKey(), propertyEntry.getValue());
+        }
+        if (!headers.containsKey("Content-Type")) {
+          conn.addRequestProperty("Content-Type", "application/json");
+        }
+
+        if (input.length() > 0) {
+          conn.setDoOutput(true);
+          try (OutputStream outputStream = conn.getOutputStream()) {
+            outputStream.write(input.getBytes());
+          }
+        }
+        responseCode = conn.getResponseCode();
+        if (responseCode >= 300) {
+          BufferedReader br = new BufferedReader(new InputStreamReader((conn.getErrorStream())));
+          StringBuilder sb = new StringBuilder();
+          String output;
+          while ((output = br.readLine()) != null) {
+            sb.append(output); }
+          exception = new IllegalStateException(String.format("Received error response. Response code: %s %s",
+                                                              responseCode, sb.toString()));
+        }
+        break;
+      } catch (MalformedURLException | ProtocolException e) {
+        throw new IllegalStateException("Error opening url connection. Reason: " + e.getMessage(), e);
+      } catch (Exception e) {
+        LOG.warn("Error making {} request to url {}",
+                 config.getObjectType(), getSinkEndpoint(config));
+        exception = e;
+      } finally {
+        if (conn != null) {
+          conn.disconnect();
+        }
+      }
+      retries++;
+    } while (retries < HubspotSinkHelper.MAX_TRIES);
```

Comment:
we can possibly add sleep between retries.
You can use awaitility, maybe add exponential sleep.